rl4co/envs/scheduling/fjsp/generator.py

```diff
@@ -0,0 +1,211 @@
+        #   1,1,1,0],
+        #   1,1,0,0]]
+        # This will be shuffled randomly to generate a machine-operation mapping
+        ma_ops_edges_unshuffled = torch.Tensor.float(
+            ma_seq_per_ops <= n_eligible_per_ops[..., None]
+        )
+        # random shuffling
+        idx = torch.rand_like(ma_ops_edges_unshuffled).argsort()
+        ma_ops_edges = ma_ops_edges_unshuffled.gather(2, idx).transpose(1, 2)
+
+        # (bs, max_ops, machines)
+        proc_times = torch.ones((bs, n_ops_max, self.num_mas))
+        proc_times = torch.randint(
+            self.min_processing_time,
+            self.max_processing_time + 1,
+            size=(bs, self.num_mas, n_ops_max),
+        )
+
+        # TODO: This should be unnecessary due to multiplication with ma-ops edges
+        # proc_times[mask[:,None].expand_as(proc_times)] = 0
+        # remove proc_times for which there is no corresponding ma-ops connection
+        proc_times = proc_times * ma_ops_edges
+        return proc_times
+
+    def _generate(self, batch_size) -> TensorDict:
+        batch_size = [batch_size] if isinstance(batch_size, int) else batch_size
+        # simulate how many operations each job has
+        n_ope_per_job = torch.randint(
+            self.min_ops_per_job,
+            self.max_ops_per_job + 1,
+            size=(*batch_size, self.num_jobs),
+        )
+
+        # determine the total number of operations per batch instance (which may differ)
+        n_ops_batch = n_ope_per_job.sum(1)  # (bs)
+        # determine the maximum total number of operations over all batch instances
+        n_ops_max = self.n_ops_max or n_ops_batch.max()
+
+        # generate a mask, specifying which operations are padded
+        pad_mask = torch.arange(n_ops_max).unsqueeze(0).expand(*batch_size, -1)
+        pad_mask = pad_mask.ge(n_ops_batch[:, None].expand_as(pad_mask))
+
+        # determine the id of the end operation for each job
+        end_op_per_job = n_ope_per_job.cumsum(1) - 1
+
+        # determine the id of the starting operation for each job
+        # (bs, num_jobs)
+        start_op_per_job = torch.cat(
+            (torch.zeros((*batch_size, 1)), end_op_per_job[:, :-1] + 1), dim=1
+        )
+
+        # here we simulate the eligible machines per operation and the processing times
+        n_eligible_per_ops = torch.randint(1, self.num_mas + 1, (*batch_size, n_ops_max))
+        n_eligible_per_ops[pad_mask] = 0
+
+        # simulate processing times for machine-operation pairs
+        # (bs, num_mas, n_ops_max)
+        proc_times = self._simulate_processing_times(n_eligible_per_ops)
+
+        td = TensorDict(
+            {
+                "start_op_per_job": start_op_per_job,
+                "end_op_per_job": end_op_per_job,
+                "proc_times": proc_times,
+                "pad_mask": pad_mask,
+            },
+            batch_size=batch_size,
+        )
+
+        return td
+
+
+class FJSPFileGenerator(Generator):
```

Comment:
Great idea! We should do something similar in routing too

Reply:
Really nice idea! We were thinking of making this `generator.py` file a data manager, including generating data, loading from the file, and save to the file. This `FileGenerator()` is a nice implementation! Let's also apply this to other environments 😁